### ISSUE TITLE
New string function to add word chunking for multi-word, multi-line text marks

### ISSUE THREAD

**Issue by atiro** [vega/vega#3963]
With multi-line strings in text marks I'm trying to find a way to be able to turn a label coming from the data values like:
 
  "National Museum of Art and Charts"

into a 3 element string array like:

 ["National Museum", "of Art", "and Charts"]

so the text (in this case on a treemap) only takes up 3 lines. I can't seem to find an easy way to do this though, as the split method on the string will only produce a word per line when splitting on space. I wondered about, if there is no other way round this, a new string method being added to support string chunking, something along the lines of (forgive my JS please):

  seperator_value = ' '
  join_value=' '
  chunk_size = 2
  name = Array.from("National Museum of Art and Charts".split(seperator_value))
  name.slice(name.length/chunk_size).map((_,i)=>name.slice(i*=chunk_size,i+chunk_size).join(join_value)).filter((line) =>  line != "" )

**Comment by domoritz** [comment]
Hmm, I want to make sure we have something generic. for very specific methods, you can always add them at runtime. 

But let's see whether we can have maybe generic functions you can compose. So we already have split and length. So maybe what you want is an argument to the split method that says how many elements should be in a chunk.

Or we could introduce an argument for the maximum length of each element. `chunk_words('...', 10)` that does https://stackoverflow.com/a/57023599/214950.

**Comment by atiro** [comment]
Can that extra chunk argument be added to split without needing to give a limit argument as well? (or need to set -1 or such for all elements ?)

The more generic functions I can think of would all involve some form of map/splice functions to use with join/split, which seems a lot to add.

**Comment by domoritz** [comment]
There would be a default so we don't break existing uses. 

Map and split could be reasonable to support. How would you propose it?

**Comment by atiro** [comment]
I guess from a JS one-liner:

  name.split(' ').map((_, i) => { return name.split(' ').splice(i*2, 2).join(' ') }).slice(0,Math.ceil(name.split(' ').length/2))

trying to translate that into Vega Expression language, perhaps something like ( I assume the map call needs to be broken out into parts for naming the element/index and the function to call):

  slice( map(name.split(' '), "_", "i", "join(name.split(' ').splice(i*2, 2), ' ')" ), 0, ceil(name.split(' ').length/2)

**Comment by atiro** [comment]
Oh actually splice isn't even needed is it:

 slice( map(name.split(' '), "_", "i", "join(slice(name.split(' '), i*2, i*2 + 2), ' ')" ), 0, ceil(name.split(' ').length/2)

**Comment by domoritz** [comment]
Oh so it already works? If yes, we should start with that and see what minimal changes would make it easier/more convenient.

**Comment by atiro** [comment]
Would work with the addition a map function for arrays that could have a expression called on each element (with access to element value and array index) to return a new array. Seems like that could be useful for other things as well?

**Comment by domoritz** [comment]
Yes. Feel free to take a stab at implementing something along those lines.
